src/typescript-worker/types-fetcher.ts

```diff
@@ -0,0 +1,291 @@
+    const dtsPath = changeFileExtension(jsPath, 'd.ts');
+    const dtsSpecifier = `${pkg}/${dtsPath}`;
+    if (this._handledSpecifiers.has(dtsSpecifier)) {
+      return;
+    }
+    this._handledSpecifiers.add(dtsSpecifier);
+    const dtsResult = await this._fetchAsset(dtsSpecifier);
+    if (dtsResult.error !== undefined) {
+      return;
+    }
+    await this._handleBareAndRelativeSpecifiers(dtsResult.result, {
+      pkg,
+      path: jsPath,
+    });
+  }
+
+  private _resolveBareModule(bare: string): string {
+    const result = this._moduleResolver.resolve(bare, '');
+    if (result.type !== 'bare') {
+      throw new Error(`Only expected bare module for specifier ${bare}`);
+    }
+    return result.url;
+  }
+
+  private async _handleRelativeSpecifier(
+    relative: string,
+    referrerSpecifier: NodePackage
+  ): Promise<void> {
+    const ext = fileExtension(relative);
+    if (ext === '') {
+      // No extension is presumed js.
+      relative += '.js';
+    } else if (ext !== 'js') {
+      // Unhandled kind of import.
+      return;
+    }
+    // Note the base URL doesn't matter here, we're just using the URL API to
+    // perform path resolution.
+    const jsPath = new URL(
+      relative,
+      new URL(referrerSpecifier.path, 'https://example.com')
+    ).pathname.slice(1); // Remove the leading '/'.
+    const dtsPath = changeFileExtension(jsPath, 'd.ts');
+    const dtsSpecifier = `${referrerSpecifier.pkg}/${dtsPath}`;
+    if (this._handledSpecifiers.has(dtsSpecifier)) {
+      return;
+    }
+    this._handledSpecifiers.add(dtsSpecifier);
+    const dtsResult = await this._fetchAsset(dtsSpecifier);
+    if (dtsResult.error !== undefined) {
+      return;
+    }
+    await this._handleBareAndRelativeSpecifiers(dtsResult.result, {
+      pkg: referrerSpecifier.pkg,
+      path: jsPath,
+    });
+  }
+
+  private async _fetchAsset(
+    specifier: string
+  ): Promise<Result<string, number>> {
+    let deferred = this._specifierToFetchResult.get(specifier);
+    if (deferred !== undefined) {
+      return deferred.promise;
+    }
+    deferred = new Deferred();
+    this._specifierToFetchResult.set(specifier, deferred);
+
+    let url = this._resolveBareModule(specifier);
+    // It's common to have a module resolver that resolves to unpkg.com URLs
+    // with the ?module parameter. But ?module mode errors on .d.ts files and
+    // package.json files (anything other than .js and .html), so we have to
+    // remove that here.
+    const urlObj = new URL(url);
+    if (urlObj.host === 'unpkg.com' && urlObj.searchParams.has('module')) {
+      urlObj.searchParams.delete('module');
+      url = urlObj.href;
+    }
+    const resp = await fetch(url);
+    let result: Result<string, number>;
+    if (resp.status === 200) {
+      result = {result: await resp.text()};
+    } else {
+      result = {error: resp.status};
+    }
+    deferred.resolve(result);
+    return result;
+  }
+
+  private async _fetchPackageJson(
```

Comment:
do you want to cache these? Are you kick off multiple fetches for package.json on `lit`, `lit/decorators.js`, `lit/directives/repeat.js`, etc?

Reply:
Yeah, it is actually cached via `_fetchAsset`